### PR TITLE
Add inspection module to replace print_routes

### LINKED ISSUE #1435: Improve falcon-print-routes tool
`falcon-print-routes` is a simple (albeit somewhat spartan) tool that may come handy to list API routes.
We could polish it, and expose more widely:
* Advertise it in the documentation (currently it is only mentioned in the `1.1.0` changelog)
* Make it aware of route suffixes, a new feature in 2.0
* Expose interface to be able to list routes programmatically (suggested by @CaselIT )
* Make it clear that the tool only works with the standard `CompiledRouter` (suggested by @CaselIT )
* Anything else? (suggestions welcome!)

### PR DESCRIPTION
**Pull request by CaselIT**
# Summary of Changes

Create an inspection module to replace the print_route script.
This was inspired by the discussion about #1435 but has evolved in a inspection module that can obtain information regarding the application and create a pretty print of it.

The public interface of the inspection module has different functions that return information regarding an application.
There is a main function to inspect the entire application (routes, middleware, static_routes, sinks and error_handlers) and also individual functions for each.
The returned information are instance of classes (or list of them) which can return a pretty print, since this is more extensible than just a string.
There is also the ability to register inspection of different route class implementations by using a decorator on a inspection function that handles a particular type of router.

This is the output of a sample application
```
Falcon App (WSGI)
• Routes:
    ⇒ /foo - MyResponder:
       ├── DELETE - on_delete
       ├── GET - on_get
       └── POST - on_post
    ⇒ /foo/{id} - MyResponder:
       ├── DELETE - on_delete_id
       ├── GET - on_get_id
       └── POST - on_post_id
    ⇒ /bar - OtherResponder:
       ├── DELETE - on_delete_id
       ├── GET - on_get_id
       └── POST - on_post_id
• Middleware (Middleware are independent):
    → MyMiddleware.process_request

      ↣ MyMiddleware.process_resource
        ↣ OtherMiddleware.process_resource

            ├── Process responder

        ↢ OtherMiddleware.process_response
      ↢ CORSMiddleware.process_response
• Static routes:
    ↦ /tests/ C:\Users\Federico\Dev\GitHub\falcon\tests [C:\Users\Federico\Dev\GitHub\falcon\tests\conftest.py]
    ↦ /falcon/ C:\Users\Federico\Dev\GitHub\falcon\falcon
• Sinks:
    ⇥ /sink_cls SinkClass
    ⇥ /sink_fn sinkFn
• Error handlers:
    ⇜ RuntimeError my_runtime_handler
```

A more verbose output is

<details>
<summary>View</summary>

```
Falcon App (WSGI)
• Routes:
    ⇒ /foo - MyResponder (c:\users\federico\dev\github\falcon\try_print.py:19):
       ├── DELETE - on_delete (c:\users\federico\dev\github\falcon\try_print.py:26)
       ├── GET - on_get (c:\users\federico\dev\github\falcon\try_print.py:20)
       ├── POST - on_post (c:\users\federico\dev\github\falcon\try_print.py:23)
       ├── OPTIONS - options_responder (c:\users\federico\dev\github\falcon\falcon\responders.py:84)
       ├── CONNECT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── HEAD - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PATCH - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PUT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── TRACE - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── CHECKIN - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── CHECKOUT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── COPY - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── LOCK - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── MKCOL - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── MOVE - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PROPFIND - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PROPPATCH - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── REPORT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── UNCHECKIN - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── UNLOCK - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── UPDATE - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       └── VERSION-CONTROL - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
    ⇒ /foo/{id} - MyResponder (c:\users\federico\dev\github\falcon\try_print.py:19):
       ├── DELETE - on_delete_id (c:\users\federico\dev\github\falcon\try_print.py:35)
       ├── GET - on_get_id (c:\users\federico\dev\github\falcon\try_print.py:29)
       ├── POST - on_post_id (c:\users\federico\dev\github\falcon\try_print.py:32)
       ├── OPTIONS - options_responder (c:\users\federico\dev\github\falcon\falcon\responders.py:84)
       ├── CONNECT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── HEAD - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PATCH - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PUT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── TRACE - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── CHECKIN - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── CHECKOUT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── COPY - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── LOCK - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── MKCOL - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── MOVE - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PROPFIND - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PROPPATCH - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── REPORT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── UNCHECKIN - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── UNLOCK - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── UPDATE - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       └── VERSION-CONTROL - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
    ⇒ /bar - OtherResponder (c:\users\federico\dev\github\falcon\try_print.py:39):
       ├── DELETE - on_delete_id (c:\users\federico\dev\github\falcon\try_print.py:35)
       ├── GET - on_get_id (c:\users\federico\dev\github\falcon\try_print.py:29)
       ├── POST - on_post_id (c:\users\federico\dev\github\falcon\try_print.py:32)
       ├── OPTIONS - options_responder (c:\users\federico\dev\github\falcon\falcon\responders.py:84)
       ├── CONNECT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── HEAD - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PATCH - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PUT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── TRACE - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── CHECKIN - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── CHECKOUT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── COPY - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── LOCK - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── MKCOL - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── MOVE - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PROPFIND - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── PROPPATCH - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── REPORT - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── UNCHECKIN - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── UNLOCK - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       ├── UPDATE - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
       └── VERSION-CONTROL - method_not_allowed (c:\users\federico\dev\github\falcon\falcon\responders.py:59)
• Middleware (Middleware are independent):
    → MyMiddleware.process_request

      ↣ MyMiddleware.process_resource
        ↣ OtherMiddleware.process_resource

            ├── Process responder

        ↢ OtherMiddleware.process_response
      ↢ CORSMiddleware.process_response
    - Middlewares classes:
        ↣ CORSMiddleware (c:\users\federico\dev\github\falcon\falcon\middlewares.py:2):
           └── process_response (c:\users\federico\dev\github\falcon\falcon\middlewares.py:3)
        ↣ MyMiddleware (c:\users\federico\dev\github\falcon\try_print.py:73):
           ├── process_request (c:\users\federico\dev\github\falcon\try_print.py:74)
           └── process_resource (c:\users\federico\dev\github\falcon\try_print.py:77)
        ↣ OtherMiddleware (c:\users\federico\dev\github\falcon\try_print.py:83):
           ├── process_resource (c:\users\federico\dev\github\falcon\try_print.py:87)
           └── process_response (c:\users\federico\dev\github\falcon\try_print.py:90)
• Static routes:
    ↦ /tests/ C:\Users\Federico\Dev\GitHub\falcon\tests [C:\Users\Federico\Dev\GitHub\falcon\tests\conftest.py]
    ↦ /falcon/ C:\Users\Federico\Dev\GitHub\falcon\falcon
• Sinks:
    ⇥ /sink_cls SinkClass (c:\users\federico\dev\github\falcon\try_print.py:57)
    ⇥ /sink_fn sinkFn (c:\users\federico\dev\github\falcon\try_print.py:53)
• Error handlers:
    ⇜ Exception _python_error_handler (c:\users\federico\dev\github\falcon\falcon\app.py:894)
    ⇜ HTTPError _http_error_handler (c:\users\federico\dev\github\falcon\falcon\app.py:891)
    ⇜ HTTPStatus _http_status_handler (c:\users\federico\dev\github\falcon\falcon\app.py:888)
    ⇜ RuntimeError my_runtime_handler (c:\users\federico\dev\github\falcon\try_print.py:66)
```

</details>

This is still very much still a proof of concept wip, with documentation and test missing. I wanted to open this PR to get feedback on this.

The main items that are missing are:
- verify that everything work with an async app
- tests
- documentation on it. I think it would be useful if we include the output of the inspection in the example documentation to better visualize the configuration of the example

# Related Issues

This fixes #1435

# Pull Request Checklist

This is just a reminder about the most common mistakes.  Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](https://github.com/falconry/falcon/blob/master/CONTRIBUTING.md) at least once; it will save you a few review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing to do.

- [x] Added **tests** for changed code.
- [x] Prefixed code comments with GitHub nick and an appropriate prefix.
- [x] Coding style is consistent with the rest of the framework.
- [x] Updated **documentation** for changed code.
    - [x] Added docstrings for any new classes, functions, or modules.
    - [x] Updated docstrings for any modifications to existing code.
    - [x] Added references to new classes, functions, or modules to the relevant RST file under `docs/`.
    - [x] Updated all relevant supporting documentation files under `docs/`.
    - [x] A copyright notice is included at the top of any new modules (using your own name or the name of your organization).
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html?highlight=versionadded#directive-versionadded).
- [x] Changes (and possible deprecations) have [towncrier](https://pypi.org/project/towncrier/) news fragments under `docs/_newsfragments/`. (Run `towncrier --draft` to ensure it renders correctly.)

If you have *any* questions to *any* of the points above, just **submit and ask**! This checklist is here to *help* you, not to deter you from contributing!

*PR template inspired by the attrs project.*

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -173,7 +173,7 @@ def load_description():
     entry_points={
         'console_scripts': [
             'falcon-bench = falcon.cmd.bench:main',
-            'falcon-print-routes = falcon.cmd.print_routes:main'
+            'falcon-inspect-app = falcon.cmd.inspect_app:main'
```
Comment (vytas7) [inline setup.py]: I would prefer to keep the old alias for compatibility in case anyone is accustomed to that.

Comment (CaselIT) [reply setup.py]: I changed the alias since it does more than just logging the routes, but I think that we can keep the old alias or have both.

**Comment by CaselIT** [comment]
The code and the tests should be done.
The documentation is still missing. I was thinking of adding a new documentation page with a brief explanation of the module and an example output similar to the one in this page.
Also maybe mention it in some tutorial examples, like where it's adding the roues and the one on the middleware?

**Comment by CaselIT** [comment]
I've added a new documentation file for the inspect module and updated the quickstart and tutorial files to introduce the inspect module.

I think this is ready for a review

File: setup.py [hunk 2]
```diff
@@ -174,7 +174,8 @@ def load_description():
     entry_points={
         'console_scripts': [
             'falcon-bench = falcon.cmd.bench:main',
-            'falcon-print-routes = falcon.cmd.print_routes:main'
+            'falcon-inspect-app = falcon.cmd.inspect_app:main',
+            'falcon-print-routes = falcon.cmd.inspect_app:main',
```
Comment (kgriffs) [inline setup.py]: It would be nice if the app would print out a deprecation message when using the old script name.

Comment (CaselIT) [reply setup.py]: Yes I was also thinking about this. Maybe it would also select the "routes only switch by default"?

**Comment by CaselIT** [comment]
Take your time, it ended up beeing quite large :+1:

The suffixed responders are not treated differently than the normal ones, but the name of the function is correctly printed, like "on_get_with_suffix", but maybe there should be an attribute in the route method class that contains the suffix?

Also I was wondering if we should separate the internal routers from the source info print. Now both are under the verbose flag. Maybe we could add an "internal" flag?

**Comment by kgriffs** [comment]
> The suffixed responders are not treated differently than the normal ones, but the name of the function is correctly printed, like "on_get_with_suffix", but maybe there should be an attribute in the route method class that contains the suffix?

TBH I think it looks good as-is.

> Also I was wondering if we should separate the internal routers from the source info print. Now both are under the verbose flag. Maybe we could add an "internal" flag?

That would be great, actually.

**Comment by CaselIT** [comment]
@kgriffs I've split the internal and the verbose options.

File: docs/api/inspect.rst [hunk 3]
```diff
@@ -3,22 +3,26 @@
...
 
 .. code:: bash
 
-    # my_module is the module that defined tha application under the name app
-    falcon-inspect-app my_module:app
-
-The output would be:
-
-.. code::
+    # my_module exposes tha application as a variable named "app"
```
Comment (CaselIT) [inline docs/api/inspect.rst]: tha -> the

Comment (kgriffs) [reply docs/api/inspect.rst]: fixed, thx!

**Comment by kgriffs** [comment]
Hope you don't mind my tweaks... just trying to polish up the prose a bit.

File: falcon/inspect.py [hunk 4]
```diff
@@ -98,10 +101,12 @@ def inspect_static_routes(app: App) -> 'List[StaticRouteInfo]':
     """Inspects the static routes of an application.
 
     Args:
-        app (falcon.App): The application to inspect.
+        app (falcon.App): The application to inspect. Works with both
+            :class:`falcon.App` and :class:`falcon.asgi.App`.
 
     Returns:
-        List[StaticRouteInfo]: The list of static routes of the application.
+        List[StaticRouteInfo]: The list of static routes that have
```
Comment (CaselIT) [inline falcon/inspect.py]: A list? as above?

File: falcon/inspect.py [hunk 5]
```diff
@@ -114,10 +119,11 @@ def inspect_sinks(app: App) -> 'List[SinkInfo]':
     """Inspects the sinks of an application.
 
     Args:
-        app (falcon.App): The application to inspect.
+        app (falcon.App): The application to inspect. Works with both
+            :class:`falcon.App` and :class:`falcon.asgi.App`.
 
     Returns:
-        List[SinkInfo]: The list of sinks of the application.
+        List[SinkInfo]: The list of sinks used by the application.
```
Comment (CaselIT) [inline falcon/inspect.py]: A list? as above?

File: falcon/inspect.py [hunk 6]
```diff
@@ -131,10 +137,12 @@ def inspect_error_handlers(app: App) -> 'List[ErrorHandlerInfo]':
     """Inspects the error handlers of an application.
 
     Args:
-        app (falcon.App): The application to inspect.
+        app (falcon.App): The application to inspect. Works with both
+            :class:`falcon.App` and :class:`falcon.asgi.App`.
 
     Returns:
-        List[ErrorHandlerInfo]: The list of error handlers of the application.
+        List[ErrorHandlerInfo]: The list of error handlers used by the
```
Comment (CaselIT) [inline falcon/inspect.py]: A list? as above?

**Comment by CaselIT** [comment]
> Hope you don't mind my tweaks... just trying to polish up the prose a bit.

No, not at all. I'm not very good with the documentation :(

**Comment by CaselIT** [comment]
Nice!

**Pull request merged**